### PR TITLE
Implement model layer

### PR DESCRIPTION
**Pull request by olzraiti**

### CODE REVIEW

**Comment by olzraiti** [comment]
Retest this please

**Comment by olzraiti** [comment]
I replaced all java Iterators with for loops for better readability. I think this was a bad idea because it is unsafe to delete from a list you are iterating with a for loop. Iterator is fail safe in this sense. I will replace the for loops before merging.

File: App/build.gradle [hunk 1]
```diff
@@ -92,6 +94,10 @@ dependencies {
 
     compile 'com.android.support:support-annotations:21.0.3'
 
+    compile 'com.fasterxml.jackson.core:jackson-core:2.3.1'
```
Comment (ffagerho) [inline App/build.gradle]: Any reason not to use 2.6.1?

File: App/src/main/java/cc/softwarefactory/lokki/android/activities/SearchActivity.java [hunk 2]
```diff
@@ -173,37 +176,24 @@ protected void onPostExecute(String query) {
...
+        MainApplication.Contacts contacts = MainApplication.contacts;
+        for (String id : MainApplication.dashboard.getUserIdsICanSee()) {
+            String email = MainApplication.dashboard.getEmailByUserId(id);
+            Contact contact = (contacts != null)? contacts.getContactByEmail(email) : null;
+            String name = "";
+
+            if (contact != null) {
+                name = contact.getName();
             }
+            if (email.toLowerCase().contains(query) || name.toLowerCase().contains(query)) {
```
Comment (ffagerho) [inline App/src/main/java/cc/softwarefactory/lokki/android/activities/SearchActivity.java]: use toLowerCase on query also

File: App/src/main/java/cc/softwarefactory/lokki/android/activities/SearchActivity.java [hunk 3]
```diff
@@ -173,37 +176,24 @@ protected void onPostExecute(String query) {
...
+                //Display either name or email depending on whether a name exists
+                //Store contact data in the result's extra data for easy access
+                if(!name.isEmpty())
```
Comment (ffagerho) [inline App/src/main/java/cc/softwarefactory/lokki/android/activities/SearchActivity.java]: Style/code safety: always use braces for future-proofing

File: App/src/main/java/cc/softwarefactory/lokki/android/utilities/DialogUtils.java [hunk 4]
```diff
@@ -93,7 +93,7 @@ public void onClick(View view) {
                                         context.getString(R.string.analytics_label_confirm_name_new_place_dialog_successful));
                                 try {
                                     ServerApi.addPlace(context, value.toString(), latLng, radius);
-                                } catch (JSONException e) {
+                                } catch (Exception e) {
```
Comment (sandeepgithubrepo) [inline App/src/main/java/cc/softwarefactory/lokki/android/utilities/DialogUtils.java]: catch more specific exception and it is good idea to log the exception

File: App/src/main/java/cc/softwarefactory/lokki/android/utilities/Utils.java [hunk 5]
```diff
@@ -75,10 +77,10 @@ private static Boolean loadContacts(Context context) {
...
         }
         try {
-            MainApplication.contacts = new JSONObject(jsonData);
-            MainApplication.mapping = MainApplication.contacts.getJSONObject("mapping");
-        } catch (JSONException e) {
-            MainApplication.contacts = new JSONObject();
+            MainApplication.contacts = JSONModel.createFromJson(jsonData, MainApplication.Contacts.class);
+        } catch (IOException e) {
+            MainApplication.contacts = new MainApplication.Contacts();
+            e.printStackTrace();
```
Comment (sandeepgithubrepo) [inline App/src/main/java/cc/softwarefactory/lokki/android/utilities/Utils.java]: good to log the exception

File: App/src/test/java/cc/softwarefactory/lokki/android/espresso/SearchScreenTest.java [hunk 6]
```diff
@@ -77,10 +74,10 @@ private void forcePlacesLoad() {
      * Helper method to ensure that places are loaded before searching, otherwise tests may fail randomly
      * @throws InterruptedException
      */
-    private void waitForPlaces() throws InterruptedException {
+    private void waitForPlaces() throws InterruptedException, JsonProcessingException {
         int counter = 0;    //If we still don't have places after 1 second, let the test fail
         forcePlacesLoad();
-        while (counter < 10 && MainApplication.places.toString().equals("{}")){
+        while (counter < 10 && MainApplication.places.serialize().equals("{}")){
```
Comment (sandeepgithubrepo) [inline App/src/test/java/cc/softwarefactory/lokki/android/espresso/SearchScreenTest.java]: good idea to define static constant to hold value 10

File: App/src/main/java/cc/softwarefactory/lokki/android/services/LocationService.java [hunk 7]
```diff
@@ -337,72 +336,52 @@ private void showArrivalNotification() {
...
-                    if(placeBuzz == null || placeBuzz.getInt("buzzcount") <= 0) {
-                        break;
-                    }
+                while (buzzPlace != null && buzzPlace.getBuzzCount() > 0) {
                     Log.d(TAG, "Vibrating...");
                     Vibrator v = (Vibrator) getApplicationContext().getSystemService(Context.VIBRATOR_SERVICE);
                     v.vibrate(1000);
-                    Thread.sleep(2500);
-                    placeBuzz.put("buzzcount", placeBuzz.getInt("buzzcount") - 1);
+                        Thread.sleep(2500);
```
Comment (ffagerho) [inline App/src/main/java/cc/softwarefactory/lokki/android/services/LocationService.java]: Remove extra tab

**Comment by olzraiti** [comment]
This commit solves the issues raised in code review.

**Pull request merged**